Add Heroku button to generated app's README

gjaldon opened the pull request:
Closes https://github.com/phoenixframework/phoenix/issues/971

Apart from adding files that are needed by the Heroku deploy button, I also removed `prod.secret.exs` in favor of using ENV variables inside `prod.exs`. Should we expose a `random_string` function so users have an easy way to generate their `SECRET_KEY_BASE` for prod? I'm thinking yes, but not sure which module to put it in. 

Looking forward to your feedback!

josevalim commented:
Awesome!

The idea though is to not make Phoenix exclusive to Heroku. For example, we should still keep the database configuration and `config/prod.secret.exs` as before. That will hurt running Heroku out of the box but we could have a small set of instructions:

1. Remove `config/prod.secret.exs` from .gitignore

2. Replace the important bits by environment variables

3. Deploy!

If we could support both at the same time even better. For example, we could have `url: System.get_env(...)` by default in production as long as it does not affect the regular code.

josevalim commented on installer/templates/new/app.json:
```diff
@@ -0,0 +1,19 @@
+    "postdeploy": "mix ecto.migrate"
+  },<% end %>
+  "env": {
+    "BUILDPACK_URL": "https://github.com/ddollar/heroku-buildpack-multi",
```
@tsloughter do we have support for multiple buildpacks in app.json too?

tsloughter replied:
Argh, I just asked and you can't yet.

josevalim replied:
Do you know if there is a timeline? Who can we bother? :P

gjaldon replied:
I asked on heroku channel in freenode and a Heroku dev responded saying he had a pending PR he is waiting to be reviewed. Forgot his name though. :D So there's a pending PR but he said no definite date on when it will be shipped.

tsloughter replied:
There is a PR. Still has to be tested more.

tsloughter replied:
@gjaldon was it Owen? He is the one handling it.

tsloughter replied:
May be done this week.

gjaldon replied:
Ah yes, it was Owen. :) Thanks for the update!

tsloughter replied:
It is out: https://devcenter.heroku.com/articles/app-json-schema#buildpacks

gjaldon commented:
Was initially thinking of removing `config/prod.secret.exs` and users who don't deploy to Heroku set the DATABASE_URL and SECRET_KEY_BASE env vars in their prod env. But I guess, copying `prod.secret.exs` to their prod is more convenient? 

Anyway, will try to make it more convenient for user to deploy on Heroku or otherwise.

gjaldon commented:
@josevalim - this now provides defaults for secret_key and database_url through DATABASE_URL and SECRET_KEY_BASE env vars. Config for those will be overridden though if user copies prod.secret.exs to their prod server.

For heroku deploy button to work, user needs to comment out `import_config "prod.secret.exs"` in prod.exs. Let me know if there's anything else that needs done here.

josevalim commented:
@gjaldon Sorry for not being clear. I don't think we should have the configuration twice. I was saying we should do:

    config :foo, Foo.Repo,
      url: System.get_env("DATABASE_URL"),
      other_options

This will be fine as long as we guarantee the ones in the URL in (which I am not sure they do). :)

gjaldon commented:
@josevalim updated the instructions and removed duplicate config. I also removed multi-buildpack and used the newly implemented buidpacks opt in app.json.

josevalim commented on installer/lib/phoenix_new.ex:
```diff
@@ -226,6 +227,7 @@ defmodule Mix.Tasks.Phoenix.New do
       # Configure your database
       config :#{binding[:application_name]}, #{binding[:application_module]}.Repo,
         adapter: #{inspect binding[:adapter_module]},
+        url: System.get_env("DATABASE_URL"),
```
Cool! Have you checked if the URL configuration overrides the regular one?

gjaldon replied:
I checked and found that if url is nil, the regular ones will be used. If url is present and is valid, it will override the regular one.

josevalim commented on installer/templates/new/app.json:
```diff
@@ -0,0 +1,22 @@
+    "postdeploy": "mix ecto.migrate"
+  },<% end %>
+  "env": {
+    "SECRET_KEY_BASE": {
```
Should we ask users to set this somehow?

gjaldon replied:
This automatically generates a random secret and adds it to Heroku as a SECRET_KEY_BASE. User doesn't need to set it. I added a `System.get_env("SECRET_KEY_BASE")` to prod.secret.exs. 

By the way, also tested this from this https://github.com/gjaldon/test. I also set it to private and added a template parameter and it worked fine too.

henrik commented on installer/templates/new/README.md:
```diff
@@ -6,3 +6,16 @@ To start your new Phoenix application:
+
+[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy)
+
+The deploy button above allows you to instantly deploy your applications to Heroku from your Github
```
Github -> GitHub I believe.
